Graph meter coloring & dynamic scaling!

Explorer09 opened the pull request:
**This branch needs rebasing and it's code needs some polishing.**

This pull request is much bigger than #355 because it's improved and supports dynamic scaling. :)

Screenshot:
![Screenshot](https://cloud.githubusercontent.com/assets/2487263/15004285/bf4084b8-11e7-11e6-95c1-504abac4597c.png)

Meters that will use what I called "dynamic scaling":
- Load meter & Load average meter
- Tasks meter (a.k.a. Task counter)
- Uptime meter

Caveats:
- When merging, it's better to review changes in every commit as some of them changes variable naming in existing code for some reasons (that you might disagree; please tell me).
- The colors array in GraphData will take larger size than necessary. Although I have optimized the (tree) data structure so that it allocated no more "cells" than needed, the cell size is still large and not optimized yet. The color cells uses `ColorElements` enum type (which is sizeof(int) by default), and I didn't yet choose a way to optimize it. I'm asking suggestions about what to do here in code.*

*: There are two way of optimization in code:
(1) changing ColorElements typedef from `enum` to `short` or `char`, or
(2) use `__attribute__((packed))` compiler directive on the enum (which will require compiler to support GCC extension), or
(3) just don't bother at all.

hishamhm commented:
Oh wow, this will take me a while to go through! But I'm impressed :)

eworm-de commented:
Any chance we see this being merged any time soon?
I really like the graph meter coloring but I do not like to resolve merge conflicts again and again (or wait for @Explorer09 to do it).

Explorer09 commented:
@eworm-de 
Hi, I was not maintaining this branch for a while, partly because @hishamhm [was not interested in merging](https://github.com/hishamhm/htop/pull/651#issuecomment-318154184), and because I have been troubled about variable naming for a while.
I would appreciate if someone likes to help making this PR branch up to date.

Explorer09 committed:
GraphData constructor and destructor.

Prerequisite commit for graph meter coloring and dynamic scaling
feature.

Constuctor is a simple xCalloc(), and destructor is a simple free()
call.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Introduce GRAPH_NUM_RECORDS macro constant.

And decouple its semantics from METER_BUFFER_LEN.
GRAPH_NUM_RECORDS is not yet tweakable at runtime for now.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Rename Meter.total variable to Meter.full

The word "total" has a dictionary meaning of "sum" or "a quantity
obtained by addition of a group of members" [WordNet]. Although this
term works when we talk about "total CPU time" or "total memory", it
may be inadequate to the context of meter drawing. For example,
TasksMeter fills this variable with total number of processes; while
this fits the dictionary meaning of "total", it will go weird when it
is plotted as a graph. (And it will confuse my graph coloring
algorithm, too.)

Rename this variable to "full" to better reflect it's use - indicating
the "full" value of the bar/graph.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Rename some local variables in GraphMeterMode_draw()

Rename "i" to "index". ("i" is not a loop counter here and so worth
renaming.)
Rename "k" to "col".

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Introduce file-scope constant array "Meter_prefixes"

And replace inline string literal with it in Meter_humanUnit().

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Graph meter coloring & dynamic scaling. (Commit 1/4)

This patch adds two features to the graph meter:
1. Coloring. The colors will base on items of the (stacked) bars. Due
   to low resolution of the character terminal, the rasterized colors
   may not look nice, but it's better than nothing. :)
2. Dynamic scaling. Non-percentage graph can be drawn for meters now.
   Very useful for meters such as LoadAverage and Tasks that have no
   fixed maximum. To use non-percent graph, set a negative "full"
   variable (formerly known as "total") to the meter.

The BarMeterMode is also adapted to support negative "full" variable
now.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Graph meter coloring & dynamic scaling. (Commit 2/4)

Linear time algorithm O(GRAPH_HEIGHT + items) for color determination
(rasterization).

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Graph meter coloring & dynamic scaling. (Commit 3/4)

Improve worst case performance of GraphData_getColor()

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Graph meter coloring & dynamic scaling. (Commit 4/4)

New display option: align_graph_meter
"Align graph meter bars to reducing color blinking"
(enabled by default)

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Enable non-percent graph for Load average meters.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
TasksMeter behavior changes.

1. Enable non-percent graph display.
2. Kernel threads are now always drawn in bar and graph modes,
   regardless of htop's hideKernelThreads option. (To really remove
   kernel threads from the count it should so do by changing the
   counting on pl->totalTasks and not on the meter. I.e. The behavior
   had been incorrect until this commit.)
3. Supplemental text in bar mode now shows "running tasks / total
   tasks". Previous behavior was "running tasks / largest total tasks
   ever recorded during lifetime of the meter".

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Enable non-percent graph for Uptime meter.

Also change one behavior, that in bar or graph mode, Uptime meter will
now draw fractions of days. (Not a useful tweak though. :) )

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
GraphData_getColor() optimization

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>
